src/UI/SlotArea.h

```diff
@@ -397,14 +397,19 @@ class cSlotAreaEnchanting:
 	virtual void OnPlayerAdded  (cPlayer & a_Player) override;
 	virtual void OnPlayerRemoved(cPlayer & a_Player) override;
 
-	/* Get the count of bookshelves who stand in the near of the enchanting table */
-	int GetBookshelvesCount(cWorld & a_World);
+	/* Get the number of bookshelves which are near the enchanting table */
+	unsigned GetBookshelvesCount(cWorld & a_World);
+
+	/* Return the enchanted item matching the chosen option (0, 1, 2) */
+	cItem GetEnchantedOption(size_t a_EnchantOption);
 
 protected:
+
 	/** Handles a click in the item slot. */
 	void UpdateResult(cPlayer & a_Player);
 
 	Vector3i m_BlockPos;
+	std::array<cItem, 3> m_EnchantedItemOptions;
```

Comment:
What do you think about computing the enchantment in cClientHandle::HandleEnchant insteading of precomputing and storing?

Reply:
It's possible, but that would mean computing it twice. Once when an item is placed in the slot in order to display possible enchantment options and again when an option is clicked.

Reply:
We’d have to be careful with the PRNG state too if we did this, or reinitialise from the seed for each option in cSlotArea::update to make the re-calculation match.

Reply:
hmm, good points.

I asked because I was thinking there'd be no point in storing both the enchanted results and the window properties. We may be able to avoid storing the std::array<short, 10> stuff altogether, if we move the "sufficient experience/lapis" calculations into UpdateResult.

The client greys out unavailable enchantments, so HandleEnchant won't get them barring a hacked client. If UpdateResult does the lapis calculations, sets the available enchants into m_EnchantedItemOptions and empties the unavailable ones, then HandleEnchant can just read them

Reply:
Yep I was just about to say, storing the 3 shorts and the m_EnchantedItemOptions is probably the best. We don't need the other 7 shorts anywhere. I think it's nice to try to be secure against hacked clients where cheap and easy.

Reply:
Also I haven't done any multiplayer testing. Do you know what happens if two people were to enchant with the same table simultaneously? Do they each get their own window object?

Reply:
IIRC yes, like crafting tables and other inventories that don't store items in a block entity (and drop/return the items when the window is closed) give each player their own own window object.  Enchanting tables are just special because they *have* a block entity, but IIRC it's only used for rendering clientside and doesn't have any special behavior serverside.

Reply:
Just quickly tested and confirmed :)